smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated
 
         Returns
         -------
-        float
+        ndarray of shape (n_samples,) or (n_samples, n_features)
             Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
-        IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
         ValueError
             Input must lie in domain of surrogate.
+        NotImplementedError
+            Predict after fitting to gradient not supported.
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
-        else:
-            oob = (numpy.any(x < self.domain[0]) or 
-                    numpy.any(x > self.domain[1]))
+        # validate inputs
+        X = numpy.array(X, ndmin=2)
+        if X.shape[1] != self.num_dimensions:
+            raise IndexError("Must be 2D array with shape (n_samples, n_features)")
+        if not self._valid_cache:
+            raise RuntimeError("Model must be fit!")
+        oob = any(
+            numpy.any(X[:, i] < self.domain[i][0])
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
         if oob:
-            raise ValueError('x must lie in domain of surrogate')
-        # transform point into basis domain and evaluate
-        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
-        value = 0
-        for coeff, basis in zip(self._coefficients, self.grid.basis_functions):
-            if self.dimension > 1:
-                term = numpy.prod(
-                        [basis[i](x_scaled[...,i]) for i in range(len(basis))],
-                        axis=0)
-            else:
-                term = basis(x_scaled)
-            value += coeff*term
+            raise ValueError("X must lie in domain of surrogate")
+
+        # create lookup table
+        num_basis_max = numpy.max([len(p) for p in self._basis_sets])
```

Comment:
Is this only used for allocating the basis function calculations? If yes, I would make `lookup_table` a list of 2d NumPy arrays so that each array is only as big as is needed. This will be important for anisotropic grids, as the memory requirements will explode if one dimension has many more basis functions than another.